### ISSUE TITLE
Does masking only work for homogeneous batches?

### ISSUE THREAD

**Issue by EderSantana** [fchollet/keras#1206]
Please, let me know if I got this wrong.

I was doing some experiments with masking and it seems that it doesn't work for heterogeneous batches, I mean those where in a single batch we have sequences with different lengths. Here is why:
https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L418

It will only mask out if all the values for all dimensions and all samples are zero. This means that mask will only kick out for homogenous batches (all samples have same length). But, training with homogeneous batches don't work well since the gradient steps will be biased to each sequence length, one at time instead of properly averaging across different lengths.

In the Theano backend, sequences can have different lengths in different batches, so this kind of masking could be replaced by using batches where everybody already have similar lengths. If that is also the case for TF, we may just get rid of masking.

In case we are looking to support heterogenous batches, one solution could be:

``` python
# calculate masking per sample and reshape it to the size of the input
switch = input.sum(axis=1, keepdims=True).repeat(input.shape[1], axis=1)
output = T.switch(switch, output, 0. * output)
return_states = []
for state, new_state in zip(states, new_states):
    # same for states
    switch = input.sum(axis=1, keepdims=True).repeat(states.shape[1], axis=1)
    return_states.append(T.switch(switch, new_state, state))
```

One thing about that solutions is that we have to make sure that all states are matrices, which is the case for conventional RNN. For the case of Neural Turing Machines and others where the states are tensors, we can just reshape them back and forth.

I don't know if that solution would work for TensorFlow though. But here is one thing. Masking users should be warned to use homogenous batches for now.

Again, let me know if I got anything wrong.

**Comment by fchollet** [comment]
Conceptually I don't think other samples in the batch should affect masking for a specific sample. So that should definitely be fixed...

**Comment by fchollet** [comment]
What would be some suggestions for a fix?

Reminder of the position of the problem:

Theano:
```python
        output, new_states = step_function(input, states)
        if masking:
            # if all-zero input timestep, return
            # all-zero output and unchanged states
            switch = T.any(input)
            output = T.switch(switch, output, 0. * output)
            return_states = []
            for state, new_state in zip(states, new_states):
                return_states.append(T.switch(switch, new_state, state))
            return [output] + return_states
```

TensorFlow:
```python
        output, new_states = step_function(input, states)
        if masking:
            # for now we raise an exception because tf.reduce_any will not work
            raise Exception("Masking is Theano-only for the time being.")

            # if all-zero input timestep, return
            # all-zero output and unchanged states
            switch = tf.reduce_any(input)
            output = tf.control_flow_ops.cond(switch,
                                              lambda: output,
                                              lambda: 0. * output)
            return_states = []
            for state, new_state in zip(states, new_states):
                return_states.append(tf.control_flow_ops.cond(switch,
                                                              lambda: new_state,
                                                              lambda: state))
            states = return_states
```

**Comment by wxs** [comment]
@fchollet I'm just starting to look at this issue (we've been using a somewhat older version of Keras recently). Can  you speak to why you have returned to the "masking value" approach from the approach we had converged on before where mask was sent separately through the network? There are a host of issues with masking values that we discussed before: I don't see the appeal.

**Comment by wxs** [comment]
At any rate: an approach I would take to solving the problem would be to pass the mask into the `rnn` call, instead of using it to transform the input. Then the Theano backend can pass the mask into `scan` and use it in there like I had originally implemented before multiple backends.

To articulate the problem with the masked value approach: with ReLU activations, zero outputs are common, so there is no reason to think that it will be that uncommon to get an all-zero input to an RNN layer, especially with many stacked RNNs.

This could be further compounded by model pruning techniques which involve setting a very high percentage of weights to 0.

I can redo this backend to match how I designed masking before (with mask passed along with input), but would like to understand if there was a reason to move away from that approach before I put in the effort.

**Comment by EderSantana** [comment]
btw, I tested `Sequential` right before the `backend` and it still works with the layers defined with the new code. So, if anybody has an urgency of using the past mask and sample_weight behavior just copy this somewhere: https://github.com/fchollet/keras/blob/9c1afbb66782fe67c796fae85cd68d8f619b8aa6/keras/models.py

**Comment by fchollet** [comment]
> would like to understand if there was a reason to move away from that approach

The reason was the switch away from scan to an API that had a single input and a single output (which simplifies lots of things). The thread on the switch to the modular backend contains a discussion about the design choices made when developing the current RNN API. The reasons you mention are among those why masking is optional in the current setup, and will only be applied if a mask object is passed to the RNN layer.

If you have a better solution that involves explicit masking, then that's great.

**Comment by wxs** [comment]
OK great. I'll take a look next week — NIPS is taking up most of my time right now.

**Comment by wxs** [comment]
For reference to people looking at this issue, most of the discussion about the new design was in #1014.

My initial thoughts are that we should slightly extend the new `rnn` function to take an arbitrary mask parameter. 

`def rnn(step_function, inputs, initial_states, go_backwards=False, input_mask=None):`

It would then also need to return the mask as part of its returned tuple.

Then it's trivial to reintroduce masking via the previous approach. It does mean that when you write your `step_function` you need to be mask aware, but that is not a major issue as that was already the case.

Thoughts? I'll start building that now on my branch and we can see how it looks.

**Comment by fchollet** [comment]
@wxs what are your thoughts on https://github.com/fchollet/keras/pull/1258?

**Comment by wxs** [comment]
I think that #1258 fixes the base bug where masking "bleeds" across timesteps, but not the more subtle issue around value masking being dangerous.

Should fix the bug for most people, though the tests don't currently work because of a dependency issue. Looks right to me.

It assumes input is 3d, though the docstring implies that input can be 4d or more (which I suppose could happen with, e.g. time distributed image inputs), though I think that was already an issue with the `inputs = inputs.dimshuffle((1, 0, 2)` line.

In other words, pending tests I think it's definitely worth merging since RNNs as it stands are broken, and I will in the meantime try to build a more robust system via mask relaying. I'm intrigued by some of @dbonadiman ideas in #1014 around "apply" and "broadcast" being operations that layers can do to masks, but need to think about it some more.

**Comment by jeffzhengye** [comment]
@wxs then can you help with the test in #1258 ? Because I am not sure how to fit it into test_recurrent.py. All the test passed locally on my computer. 

Do you also have time to have a look at the bidirectional rnn BUG I mentioned in #1258 ?

**Comment by visionscaper** [comment]
@fchollet back in Dec. 2015 you accepted the pull request by @jeffzhengye that resolves this issue for Theano,[ back then there was no fix for Tensorflow](https://github.com/fchollet/keras/issues/1206#issuecomment-163352752). 

My question is, has this been resolved for Tensorflow by now? A lot of time has past, so chances are good that this has been resolved, **but I just wanted to have this confirmed**.

The issue was that masking would only be applied if all the samples in a batch had the masking value at a certain time step, instead of allowing different masking per sample, per timestep.

Thanks for your time and awesome work on Keras!
